docs/testing/yaml_schema.md

```diff
@@ -62,6 +62,7 @@ YAML schema
 |&emsp; &emsp; &emsp; hasMasksClear |list||
 |&emsp; &emsp; &emsp; notValue |NoneType,bool,int,float,list,dict|Y|
 |&emsp; &emsp; &emsp; anyOf |list||
+|&emsp; &emsp; &emsp; python |str|Y|
```

Comment:
we should not make yaml a turing complete languate

We should migrate to python code if we want python. Yaml is for simple tests.

Reply:
The Yaml runner currently uses completely unguarded `eval` to implement variable substitution. In contrast, the Python constraint evaluates the provided code snippet in a very constrained environment that does not provide access to file IO, `import`, `eval` and similar problematic functionality. I'm not claiming that it's a 100% unbreakable sandbox, but any validation snippet trying to do something nefarious is certainly going to be non-trivial and look suspicious.

I don't think the way tests are structured in the current Python test framework is very approachable, and being forced to rewrite entire test scripts as Python just to implement even moderately complex validation is frustrating and unproductive. Github Workflow YAMLs are a good model for how scripts can be broadly declarative (and hence machine readable, e.g. to automatically generate test plans. (Yes, I understand the ability to generate a test plan from a Python script exists currently, but it has various limitations and requires a list of steps that's separate and independent of the actual code implementing the steps.)

Reply:
In terms of actionable information during failures, an easy improvement would be to allow multiple constraints of the same kind (by allowing constraints to be an array), so that `A and B and C` can be written as 3 separate constraints to pinpoint any failures more precisely.

It would of course also be possible to have `assertEqual()` or similar statements within the code, but that goes against the spirit of what this PR is trying to work towards, which is that the overall structure of test remains declarative and code is used sparingly and in small snippets.

Reply:
Here is an example test script where I want to use this feature: https://github.com/project-chip/connectedhomeip/pull/34469/files#diff-e6386f2ee000b12ef29d530685caf7dd28c2d5b5d1e66f92d684166983fea61c

```
    - label: "TH reads ThreadNetworkTableSize attribute from DUT"
      command: readAttribute
      attribute: ThreadNetworkTableSize
      response:
          constraints:
              type: int8u
              python: value >= 2 * supportedFabrics
```

I think it's quite reasonable not to have to rewrite my whole test in Python to do this.

Reply:
@naveenkommareddi / @cjandhyala My personal opinions aside on this, do we know and/or can we estimate how many test plans or types of tests are not automated because of conditionals not present in YAML?

Reply:
Personally I agree with @tcarmelveilleux that I think we do need to be careful about this type of stuff, and require more review on it, but I think it's a useful tool that could unblock a lot of very simple and useful tests.

Reply:
> @naveenkommareddi / @cjandhyala @raju-apple My personal opinions aside on this, do we know and/or can we estimate how many test plans or types of tests are not automated because of conditionals not present in YAML?

@woody-apple haven't done this exercise recently but the last time we did this analysis , there were around ~20 TCs which we couldn't automate as we needed some kind of conditional checks in the steps , some of them we were able to automate once 'anyOf' support was added to the YAML framework and I believe a handful of the remaining ones we also automated them in Python.

Reply:
@ksperling-apple we have a lot of tests that can use the full python here: https://github.com/project-chip/connectedhomeip/tree/master/src/python_testing

Could we use that for tests that require very customized things rather than trying to extend YAML more and more? What is the overhead of converting an existing yaml test to python?

Reply:
We should probably figure out what improvements in the existing python tests would make it more pleasant to write and review by humans. YAML is easier to read whenever the test cases is simple. Ideally python should be somehow readable as well - the whole point of python is generally to write simple things easy. Complexity is generally our own doing.

Reply:
A Python script is never going to be as approachable or as declarative as something like YAML, unless you effectively restrict the test scripts to some sort of DSL, at which point it's no longer really full Python.

With the YAML format it's quite practical for somebody who can write an asciidoc cluster spec to also write a YAML test case, and maybe at a couple points they will get help from somebody else to write a slightly more complex validation using a Python expression. Teaching that same person to write Python tests from scratch is just not a comparable alternative.

Many of the Python tests I've look at do things like sub-class helper classes, import helper methods from various Python modules, use async/await, etc. It's just the wrong level of abstraction if you want anybody but people with deep Python experience to be able to contribute to test scripts.

Reply:
I think there's strong precedent here with how GitHub handles actions, which is what @ksperling-apple is following. This solves a problem, and makes existing blocked test cases automatable. 


@andy31415 @tcarmelveilleux Any suggestions about how to move forward here?

In my eyes of this, it:
• Adds new useful functionality to YAML that will unblock simple test cases
• Follows a pretty strong precedent for these type of workflows, GitHub workflows, for example does this, and that's deployed at scale
• Doesn't remove the ability to write Python tests.

I'm struggling to understand the issue with this?

Reply:
Marking this as resolved then, given this is stale.